assets/js/src/frontend.js

```diff
@@ -1,102 +1,38 @@
-// Keep a record of the page we have loaded, per query.
-const pages = {};
-
-/**
- * Gets the query and page from a URL string
- * @param {string} str The query string for url
- * @returns {query: number, page: number} | null
- */
-const extractQueryParams = ( str ) => {
-	// Default values
-	let query = null;
-	let page = null;
-
-	// Regular expressions to match query and page
-	const queryMatch = str.match( /query-(\d+)/ );
-	const pageMatch = str.match( /page=(\d+)/ );
-
-	// Extract and convert to integer if matches are found
-	if ( queryMatch ) {
-		query = parseInt( queryMatch[ 1 ], 10 );
-	}
-
-	if ( pageMatch ) {
-		page = parseInt( pageMatch[ 1 ], 10 );
-	}
-
-	// If either are null, return null
-	if ( query === null || page === null ) {
-		return null;
-	}
-
-	return { query, page };
-};
-
-/**
- * Checks if a given query and page is in the pages object
- *
- * @param {number} query The query ID
- * @param {number} page The page number
- *
- * @returns {boolean}
- */
-const isPageLoaded = ( query, page ) => {
-	if ( pages[ query ] && pages[ query ].includes( page ) ) {
-		return true;
-	} else {
-		return false;
-	}
-};
+import domReady from '@wordpress/dom-ready';
+
+const intersectionObserver = new window.IntersectionObserver( ( entries ) => {
+	entries.forEach( ( entry ) => {
+		// If intersectionRatio is 0, the target is out of view.
+		if ( entry.intersectionRatio <= 0 ) {
+			return;
+		}
+
+		// load posts
+		fetchPosts( entry.target );
+	} );
+} );
 
 /**
- * Logs a given query and page to the pages object
- *
- * @param {number} query The query ID
- * @param {number} page The page number
+ * Load page from server, extract and append new posts to button's query block.
  *
- * @returns {void}
+ * @param {*} button
  */
-const logPage = ( query, page ) => {
-	if ( ! pages[ query ] ) {
-		pages[ query ] = [];
+const fetchPosts = ( button ) => {
+	const url = button.href;
+	const container = button
+		.closest( '.wp-block-query' )
+		?.querySelector( '.wp-block-post-template' );
+
+	// return early if button is still loading or required data not found
+	if ( button.classList.contains( 'loading' ) || ! container || ! url ) {
```

Comment:
In an attempt to exploit the block, I tried to inject a different URL in the `<a>` tag to make the block load posts from an external source.

While a cross-origin rule blocked the attempt, it might be good to consider frontend validation around this line. Just in case the site (insecurely) allows calls to external sources.

Something like:

```
 const urlObj = new URL(url, window.location.origin);
if (urlObj.origin !== window.location.origin) {
    return;
}
```

Just a comment. Not a blocker to merge.